src/Swift.Bindings/src/Marshaler/MarshallingExtensions.cs

```diff
@@ -0,0 +1,46 @@
+
+
+using Swift.Runtime;
+
+namespace BindingsGeneration
+{
+    public static class MarshallingExtensions // TODO: Find better place for those
+    {
+        public static bool RequiresIndirectResult(this MethodDecl methodDecl, BaseDecl parentDecl, TypeDatabase typeDatabase)
+        {
+            if (methodDecl.IsConstructor && !(parentDecl is StructDecl structDecl && structDecl.IsMarshalledAsStruct())) return true;
+            var returnType = methodDecl.CSSignature.First();
+
+            if (returnType.Name == "") return false; // TODO: Void should be handled differently
+
+            var returnTypeRecord = typeDatabase.Registrar.GetType(returnType);
+            if (!(returnTypeRecord.IsFrozen && returnTypeRecord.IsBlittable)) return true;
+            return false;
+        }
+
+        public static bool RequiresSwiftSelf(this MethodDecl methodDecl, BaseDecl parentDecl)
+        {
+            if (parentDecl is ModuleDecl) return false; // global funcs
+            if (methodDecl.MethodType == MethodType.Static) return false;
+            if (methodDecl.IsConstructor) return false;
+
+            return true;
+        }
+
+        public static bool IsMarshalledAsStruct(this StructDecl decl)
+        {
+            return decl is StructDecl structDecl && structDecl.IsFrozen && structDecl.IsBlittable;
+        }
+
+        public static TypeRecord GetType(this TypeRegistrar typeRegistrar, ArgumentDecl argumentDecl)
```

Comment:
I agree. Currently type database accepts queries for types using module name and type name. I believe a lot of the queries will be driven by looking at some typeSpec, so maybe we can include that in the interface

Reply:
Here's something to put a pin in: Swift is simple in that there is only one namespace: the module name. C# not so much. In either case, it might be handy to have something like this:
```
public class ClassName { // implement IEquatable<> etc.
   public ClassName(string namespace, string name)
   {
        Namespace = namespace;
        Name = name;
        FullName = Namespace + "." + Name;
   }
   public string Namespace { get; init; }
   public string Name { get; init; } // may include inner classes as a path
   public string FullName { get; init }
}
```
And use this wherever you want to refer to either a C# class or a Swift class.
When writing code non-linearly, we can do `namespaces.AddIfNotPresent(clname.Namespace)` or `imports.AddIfNotPresent(clname.Namespace)`.